core/src/main/java/org/jclouds/http/internal/BaseHttpCommandExecutorService.java

```diff
@@ -80,6 +81,64 @@ protected BaseHttpCommandExecutorService(HttpUtils utils, ContentMetadataCodec c
       this.wire = checkNotNull(wire, "wire");
       this.idempotentMethods = ImmutableSet.copyOf(idempotentMethods.split(","));
    }
+   @Override
+   public ListenableFuture<HttpResponse> invokeAsync(final HttpCommand command){
+      SettableFuture<HttpResponse> finalFuture = SettableFuture.create();
+      this.invokeAsyncInternal(command, finalFuture);
+      return finalFuture;
+   }
+
+   private void invokeAsyncInternal(final HttpCommand command, final SettableFuture<HttpResponse> finalFuture) {
+      ListenableFuture<HttpResponse> future = null;
+      try {
+         final HttpRequest request = command.getCurrentRequest();
+         final Q nativeRequest = getNativeRequest(request);
+         future = invokeAsync(nativeRequest);
+         Futures.addCallback(future, new FutureCallback<HttpResponse>() {
+            @Override
+            public void onSuccess(HttpResponse response) {
+               logResponse(request, response);
+               int statusCode = response.getStatusCode();
+               if (statusCode >= 300) {
+                  if (shouldContinue(command, response))
+                     invokeAsyncInternal(command, finalFuture);
+                  else
+                     cleanup(nativeRequest);
+                  setResultOrException(command, finalFuture, response);
+               } else {
+                  cleanup(nativeRequest);
+                  setResultOrException(command, finalFuture, response);
+               }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+               handleInvokeAsyncFailure(t, command, finalFuture);
+            }
+         });
+      } catch (Exception e) {
+         handleInvokeAsyncFailure(e, command, finalFuture);
+
+      }
```

Comment:
@nacx In the retry handler, we are calling different methods i.e invoke in one case and invokeAsync in another case. I have to think whether we can avoid it or not. We can wrap sync in immediateFuture and I don't think it will have any performance impact as such. Are you suggesting to change the return type of invoke method to Future?